File: intellij-lsp/src/com/github/gtache/lsp/contributors/icon/LSPIconProvider.scala
package com.github.gtache.lsp.contributors.icon

import javax.swing.Icon

import com.github.gtache.lsp.ServerStatus
import com.github.gtache.lsp.client.languageserver.serverdefinition.LanguageServerDefinition
import com.intellij.openapi.extensions.ExtensionPointName
import org.eclipse.lsp4j.{CompletionItemKind, SymbolInformation, SymbolKind}

object LSPIconProvider {
  val EP_NAME: ExtensionPointName[LSPIconProvider] = ExtensionPointName.create("com.github.gtache.lsp.contributors.icon.LSPIconProvider")

  def getDefaultCompletionIcon(kind: CompletionItemKind) : Icon = LSPDefaultIconProvider.getCompletionIcon(kind)

  def getDefaultStatusIcons: Map[ServerStatus, Icon] = LSPDefaultIconProvider.getStatusIcons

  def getDefaultSymbolIcon(kind: SymbolKind) : Icon = LSPDefaultIconProvider.getSymbolIcon(kind)
}

trait LSPIconProvider {

  def getCompletionIcon(kind: CompletionItemKind): Icon = LSPDefaultIconProvider.getCompletionIcon(kind)

  def getStatusIcons : Map[ServerStatus, Icon] = LSPDefaultIconProvider.getStatusIcons

  def getSymbolIcon(kind: SymbolKind) : Icon = LSPDefaultIconProvider.getSymbolIcon(kind)

  def isSpecificFor(serverDefinition: LanguageServerDefinition) : Boolean
}

File: intellij-lsp/src/main/kotlin/com/github/gtache/lsp/contributors/icon/LSPIconProvider.kt
package com.github.gtache.lsp.contributors.icon

import com.github.gtache.lsp.client.languageserver.status.ServerStatus
import com.github.gtache.lsp.client.languageserver.serverdefinition.LanguageServerDefinition
import com.intellij.openapi.extensions.ExtensionPointName
import org.eclipse.lsp4j.CompletionItemKind
import org.eclipse.lsp4j.SymbolKind
import javax.swing.Icon


interface LSPIconProvider {

    companion object {
        val EP_NAME: ExtensionPointName<LSPIconProvider> = ExtensionPointName.create("com.github.gtache.lsp.lspIconProvider")

        fun getDefaultCompletionIcon(kind: CompletionItemKind): Icon? = LSPDefaultIconProvider.getCompletionIcon(kind)

        fun getDefaultStatusIcons(): Map<ServerStatus, Icon> = LSPDefaultIconProvider.statusIcons

        fun getDefaultSymbolIcon(kind: SymbolKind): Icon? = LSPDefaultIconProvider.getSymbolIcon(kind)
    }

    val statusIcons: Map<ServerStatus, Icon>
        get() = LSPDefaultIconProvider.statusIcons

    fun getCompletionIcon(kind: CompletionItemKind): Icon? = LSPDefaultIconProvider.getCompletionIcon(kind)

    fun getSymbolIcon(kind: SymbolKind): Icon? = LSPDefaultIconProvider.getSymbolIcon(kind)

    fun isSpecificFor(serverDefinition: LanguageServerDefinition): Boolean
}